### DISCUSSION TITLE
Webpack 5 Custom Unsafe Resolver Cache Issue with File Extension Changes

### DISCUSSION THREAD

**Discussion by jpbriggs408** [webpack/webpack#18089] [Q&A]
Hello Webpack Team,

We've encountered an issue while upgrading to Webpack 5, specifically related to our implementation of a custom object for `resolve.unsafeCache` to handle file extension changes without causing build failures.

## Background
In Webpack 4, we faced challenges with `resolve.unsafeCache`. We raised those issues with the Webpack team at the time and the feedback was received positively, resulting in an update to Webpack's documentation. See here for more context: https://github.com/webpack/webpack/issues/10771.

The TLDR is that turning it off resulted in slow reload times, while having it on caused build failures when a file was moved or renamed due to the cache pointing to a non-existent file. We found a workaround by implementing a custom cache object and passing it to `resolve.unsafeCache`, which checks for the existence of a file on a cache hit and deletes the cache entry if the file does not exist. This solution maintained performance benefits without leading to file not found errors.

## Issue
After upgrading to Webpack 5, we've continued using our custom cache object strategy. However, we're experiencing similar build failures that we had initially aimed to avoid with this approach. It seems the custom object is not functioning as intended with Webpack 5's handling of the cache.

## Example Configuration
```javascript
const fs = require('fs');

const realUnsafeCache = {};
const unsafeCacheHandler = {
  get(cache, key) {
    const cachedValue = cache[key];
    if (cachedValue && !fs.existsSync(cachedValue.path)) {
      delete cache[key];
      return undefined;
    }
    return cachedValue;
  },
};
const theProxiedCache = new Proxy(realUnsafeCache, unsafeCacheHandler);

// and in our config:
resolve: {
    unsafeCache: theProxiedCache,
}
```

## Questions
Have there been a change in Webpack 5 that could affect the behavior of a custom cache object passed to `resolve.unsafeCache`?
Are there recommended practices or alternative approaches for handling dynamic file extension changes (e.g., JS to TS) in Webpack 5 to avoid such build failures without sacrificing build performance? Our users are unable to rename files, switch branches, etc. without restarting the service.

## Previous Context
This approach was initially adopted to mitigate issues encountered in Webpack 4. Our journey and the challenges faced were discussed in these GitHub issues: [webpack#10771](https://github.com/webpack/webpack/issues/10771) and [webpack#7976 (comment)](https://github.com/webpack/webpack/issues/7976#issuecomment-820706796), which provide additional context and related discussions.

## Environment
Webpack version: 5.89.0
Node.js version: 16.13.2
Operating System: CentOS 7

We are seeking guidance on how to adapt our custom cache strategy for Webpack 5 or if there are new recommended approaches to handling this scenario. Any assistance or insights would be greatly appreciated.

Thank you for your time and support.

**Comment by alexander-akait** [comment]
Hello, we have

> Are there recommended practices or alternative approaches for handling dynamic file extension changes (e.g., JS to TS) in Webpack 5

We have the `extensionAlias` option https://github.com/webpack/enhanced-resolve?tab=readme-ov-file#resolver-options and https://webpack.js.org/configuration/resolve/#resolveextensionalias, it can help

Anyway if you create a small reproducible example I can look and give advices more deeply, thank you

**Comment by jpbriggs408** [comment]
Thank you for your suggestion to use the `extensionAlias` feature to handle dynamic file extension changes. I do believe that I have implemented the configuration as recommended, but unfortunately, the issue persists.

As requested, I've created a minimal reproducible example to demonstrate the issue. Let me know what you find, and thanks again!

https://github.com/jpbriggs408/Webpack-UnsafeCache-ModuleBuildError-Repro

**Comment by jpbriggs408** [comment]
I wanted to share an update on the Webpack 5 caching issue we've been experiencing. After thorough investigation, we discovered that our custom Proxy cache object is being overwritten. Specifically, the `removeOperations` function in Webpack is executed recursively on resolve options, converting our Proxy cache into a plain empty object. This occurs just before the Webpack options are handed off to `enhanced-resolve`.

In `enhanced-resolve`, we should be able to pass a boolean or an object to `unsafeCache`. If an object is passed, that object will then be used as the cache itself. This functionality is crucial for maintaining custom caching logic in complex projects.

As such, I have [opened an issue](https://github.com/webpack/webpack/issues/18129) with some more details. Looking forward to discussing how we can best solve this, if you are open to changing this behavior.
